doc/man/pam_pwquality.8
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.de FN
\fI\|\\$1\|\fP
..
.TH PAM_PWQUALITY 8 "16 Jul 2015" "Red Hat, Inc."
.SH NAME
pam_pwquality \- PAM module to perform password quality checking
.SH SYNOPSIS
\fBpam_pwquality\&.so\fR [\fI\&.\&.\&.\fR]
.SH DESCRIPTION
.PP
This module can be plugged into the
\fIpassword\fR
stack of a given service to provide some plug\-in strength\-checking
for passwords\&. The code was originally based on pam_cracklib module
and the module is backwards compatible with its options\&.
.PP
The action of this module is to prompt the user for a password and check
its strength against a system dictionary and a set of rules for identifying
poor choices\&.
.PP
The first action is to prompt for a single password, check its strength
and then, if it is considered strong, prompt for the password a second time
(to verify that it was typed correctly on the first occasion)\&. All being
well, the password is passed on to subsequent modules to be installed as the
new authentication token\&.
.PP
The strength checks works in the following manner: at first the
\fBCracklib\fR
routine is called to check if the password is part of a dictionary; if this
is not the case an additional set of strength checks is done\&. These checks
are:
.PP
Palindrome
.RS 4
Is the new password a palindrome?
.RE
.PP
Case Change Only
.RS 4
Is the new password the the old one with only a change of case?
.RE
.PP
Similar
.RS 4
Is the new password too much like the old one? This is primarily controlled
by one argument,
\fBdifok\fR
which is a number of character changes (inserts, removals, or replacements)
between the old and new password that are enough to accept the new
password\&.
.RE
.PP
Simple
.RS 4
Is the new password too small? This is controlled by 6 arguments
\fBminlen\fR,
\fBmaxclassrepeat\fR,
\fBdcredit\fR,
\fBucredit\fR,
\fBlcredit\fR, and
\fBocredit\fR\&. See the section on the arguments for the details of how
these work and there defaults\&.
.RE
.PP
Rotated
.RS 4
Is the new password a rotated version of the old password?
.RE
.PP
Same consecutive characters
.RS 4
Optional check for same consecutive characters\&.
.RE
.PP
Too long monotonic character sequence
.RS 4
Optional check for too long monotonic character sequence\&.
.RE
.PP
Contains user name
.RS 4
Optional check whether the password contains the user\*(Aqs name in some form\&.
.RE
.PP
These checks are configurable either by use of the module arguments
or by modifying the \fB/etc/security/pwquality.conf\fR configuration file\&. The
module arguments override the settings in the configuration file\&.
.PD
.SH OPTIONS
.PP
\fBdebug\fR
.RS 4
This option makes the module write information to
\fBsyslog\fR(3)
indicating the behavior of the module (this option does not write password
information to the log file)\&.
.RE
.PP
\fBauthtok_type=\fR\fB\fIXXX\fR\fR
.RS 4
The default action is for the module to use the following prompts when
requesting passwords: "New UNIX password: " and
"Retype UNIX password: "\&. The example word
\fIUNIX\fR
can be replaced with this option, by default it is empty\&.
.RE
.PP
\fBretry=\fR\fB\fIN\fR\fR
.RS 4
Prompt user at most
\fIN\fR
times before returning with error\&. The default is
\fI1\fR\&.
.RE
.PP
\fBdifok=\fR\fB\fIN\fR\fR
.RS 4
This argument will change the default of
\fI1\fR
for the number of changes in the new password from the old password\&.
.RE
.PP
\fBminlen=\fR\fB\fIN\fR\fR
.RS 4
The minimum acceptable size for the new password (plus one if credits are not
disabled which is the default)\&. In addition to the number of characters in
the new password, credit (of +1 in length) is given for each different kind
of character (\fIother\fR,
\fIupper\fR,
\fIlower\fR
and
\fIdigit\fR)\&. The default for this parameter is
\fI8\fR
\&. Note that there is a pair of length limits also in
\fICracklib\fR,
which is used for dictionary checking, a "way too short" limit of 4 which
is hard coded in and a build time defined limit (6) that will be checked
without reference to \fBminlen\fR\&.
.RE
.PP
\fBdcredit=\fR\fB\fIN\fR\fR
.RS 4
(N >= 0) This is the maximum credit for having digits in the new password\&.
If you have less than or
\fIN\fR
digits, each digit will count +1 towards meeting the current
\fBminlen\fR
value\&. The default for
\fBdcredit\fR
is
\fI0\fR
which means there is no bonus for digits in password\&.
.sp
(N < 0) This is the minimum number of digits that must be met for a new
password\&.
.RE
.PP
\fBucredit=\fR\fB\fIN\fR\fR
.RS 4
(N >= 0) This is the maximum credit for having upper case letters in the
new password\&. If you have less than or
\fIN\fR
upper case letters each letter will count +1 towards meeting the current
\fBminlen\fR
value\&. The default for
\fBucredit\fR
is
\fI0\fR
which means there is no bonus for upper case letters in password\&.
.sp
(N < 0) This is the minimum number of upper case letters that must be met
for a new password\&.
.RE
.PP
\fBlcredit=\fR\fB\fIN\fR\fR
.RS 4
(N >= 0) This is the maximum credit for having lower case letters in the
new password\&. If you have less than or
\fIN\fR
lower case letters, each letter will count +1 towards meeting the current
\fBminlen\fR
value\&. The default for
is
\fI0\fR
which means there is no bonus for lower case letters in password\&.
.sp
(N < 0) This is the minimum number of lower case letters that must be met
for a new password\&.
.RE
.PP
\fBocredit=\fR\fB\fIN\fR\fR
.RS 4
(N >= 0) This is the maximum credit for having other characters in the new
password\&. If you have less than or
\fIN\fR
other characters, each character will count +1 towards meeting the current
\fBminlen\fR
value\&. The default for
\fBocredit\fR
is
\fI0\fR
which means there is no bonus for other characters in password\&.
.sp
(N < 0) This is the minimum number of other characters that must be met for
a new password\&.
.RE
.PP
\fBminclass=\fR\fB\fIN\fR\fR
.RS 4
The minimum number of required classes of characters for the new password\&.
The default number is zero\&. The four classes are digits, upper and lower
letters and other characters\&. The difference to the
\fBcredit\fR
check is that a specific class if of characters is not required\&. Instead
\fIN\fR
out of four of the classes are required\&.
.RE
.PP
\fBmaxrepeat=\fR\fB\fIN\fR\fR
.RS 4
Reject passwords which contain more than N same consecutive characters\&.
The default is 0 which means that this check is disabled\&.
.RE
.PP
\fBmaxsequence=\fR\fB\fIN\fR\fR
.RS 4
Reject passwords which contain monotonic character sequences longer than N\&.
The default is 0 which means that this check is disabled\&.
Examples of such sequence are \*(Aq12345\*(Aq or \*(Aqfedcb\*(Aq\&. Note that
most such passwords will not pass the simplicity check unless the sequence
is only a minor part of the password\&.
.RE
.PP
\fBmaxclassrepeat=\fR\fB\fIN\fR\fR
.RS 4
Reject passwords which contain more than N consecutive characters of the
same class\&.
The default is 0 which means that this check is disabled\&.
.RE
.PP
\fBgecoscheck=\fR\fB\fIN\fR\fR
.RS 4
If nonzero, check whether the individual words longer than 3 characters
from the
\fBpasswd GECOS\fR
field of the user are contained in the new password\&.
The default is 0 which means that this check is disabled\&.
.RE
.PP
\fBdictcheck=\fR\fB\fIN\fR\fR
.RS 4
If nonzero, check whether the password (with possible modifications)
matches a word in a dictionary\&. Currently the dictionary check is performed
using the
\fBcracklib\fR
library\&.
The default is 1 which means that this check is enabled\&.
.RE
.PP
\fBbadwords=\fR\fB\fI<list of words>\fR\fR
.RS 4
The words more than 3 characters long from this space separated list are
individually searched for and forbidden in the new password\&.
By default the list is empty which means that this check is disabled\&.
.RE
.PP
\fBenforce_for_root\fR
.RS 4
The module will return error on failed check even if the user changing the
password is root\&. This option is off by default which means that just
the message about the failed check is printed but root can change
the password anyway\&. Note that root is not asked for an old password
so the checks that compare the old and new password are not performed\&.
.RE
.PP
\fBlocal_users_only\fR
.RS 4
The module will not test the password quality for users that are not present
in the \fI/etc/passwd\fR file\&. The module still asks for the password so
the following modules in the stack can use the \fBuse_authtok\fR option\&.
This option is off by default\&.
.RE
.PP
\fBuse_authtok\fR
.RS 4
This argument is used to
\fIforce\fR
the module to not prompt the user for a new password but use the one
provided by the previously stacked
\fIpassword\fR
module\&.
.RE
.PP
\fBdictpath=\fR\fB\fI/path/to/dict\fR\fR
.RS 4
Path to the cracklib dictionaries\&.
.RE

.PD
.SH "MODULE TYPES PROVIDED"
.PP
Only the
\fBpassword\fR
module type is provided\&.

.PD
.SH "RETURN VALUES"
.PP
.PP
PAM_SUCCESS
.RS 4
The new password passes all checks\&.
.RE
.PP
PAM_AUTHTOK_ERR
.RS 4
No new password was entered, the username could not be determined or the
new password fails the strength checks\&.
.RE
.PP
PAM_AUTHTOK_RECOVERY_ERR
.RS 4
The old password was not supplied by a previous stacked module or got not
requested from the user\&. The first error can happen if
\fBuse_authtok\fR
is specified\&.
.RE
.PP
PAM_SERVICE_ERR
.RS 4
A internal error occurred\&.
.RE
.SH "EXAMPLES"
.PP
For an example of the use of this module, we show how it may be stacked with the password component of
\fBpam_unix\fR(8)
.sp
.if n \{\
.RS 4
.\}
.nf
#
# These lines stack two password type modules\&. In this example the
# user is given 3 opportunities to enter a strong password\&. The
# "use_authtok" argument ensures that the pam_unix module does not
# prompt for a password, but instead uses the one provided by
# pam_pwquality\&.
#
passwd  password required       pam_pwquality\&.so retry=3
passwd  password required       pam_unix\&.so use_authtok

.fi
.if n \{\
.RE
.\}
.PP
Another example (in the
/etc/pam\&.d/passwd
format) is for the case that you want to use md5 password encryption:
.sp
.if n \{\
.RS 4
.\}
.nf
#%PAM\-1\&.0
#
# These lines allow a md5 systems to support passwords of at least 14
# bytes with extra credit of 2 for digits and 2 for others the new
# password must have at least three bytes that are not present in the
# old password
#
password  required pam_pwquality\&.so \e
               difok=3 minlen=15 dcredit= 2 ocredit=2
password  required pam_unix\&.so use_authtok nullok md5

.fi
.if n \{\
.RE
.\}
.PP
And here is another example in case you don\'t want to use credits:
.sp
.if n \{\
.RS 4
.\}
.nf
#%PAM\-1\&.0
#
# These lines require the user to select a password with a minimum
# length of 8 and with at least 1 digit number, 1 upper case letter,
# and 1 other character
#
password  required pam_pwquality\&.so \e
               dcredit=\-1 ucredit=\-1 ocredit=\-1 lcredit=0 minlen=8
password  required pam_unix\&.so use_authtok nullok md5
.fi
.if n \{\
.RE
.\}
.sp
.PD
.SH "SEE ALSO"
pwscore(1), pwquality.conf(5), pam_pwquality(8),
pam.conf(5), PAM(8)

.SH AUTHORS
.nf
Tomas Mraz <tmraz@redhat\&.com>
Original author of pam_cracklib module Cristian Gafton <gafton@redhat\&.com>
.fi


NEWS
libpwquality NEWS -- history of user-visible changes.

Release 1.2.4
* Add Python3 module subpackage

Release 1.2.3
* Fix problem with parsing the pam_pwquality options
  patch by Vladimir Sorokin.
* Updated translations from Transifex
* Treat empty user or password as NULL

Release 1.2.2

* Manual page fixes.
* Make it possible to set the maxsequence configuration value.
* Updated translations from Transifex.

Release 1.2.1

* Properly free pwquality settings.
* Add extern "C" to public header.
* Updated translations from Transifex.

Release 1.2.0

* Add maxsequence check for too long monotonic character sequence.
* Clarified alternative licensing to GPLv2+.
* Add local_users_only option to skip the pwquality checks for
  non-locals. (thanks to Stef Walter)

Release 1.1.1

* Use rpm built-in filtering of provides (fix for rhbz#830153)
* Remove strain debug fprintf() (fix for rhbz#831567)
* Make the Python bindings build optional (thanks to Colin Walters)
* Make the PAM module build optional (thanks to Jasper Lievisse Adriaanse)

Release 1.1.0

* Fixed a memory leak when throwing PWQError exception
* Added pkgconfig file (thanks to Matthias Classen)
* The simplicity checks are now called before the cracklib check
* Added enforce_for_root option to the PAM module
* Updated translations from Transifex

Release 1.0.0

* Added a check for words from user's GECOS
* Added a check for configurable words
* Added a check for maximum consecutive characters of the same class
* Fixed configuration file parsing (allowed '=' but not required)
* Fixed possible leak when string setting is set
* Project added to Transifex for localization

Release 0.9.9

* Release candidate for 1.0
* Added Python bindings
* Added user name parameter to the pwquality_check()
* Added manpages and other documentation
* Removed obsolete and unused difignore option

Release 0.9

* First prerelease


src/pam_pwquality.c
/*
 * PAM module for password quality checking using libpwquality
 *
 * See the end of the file for Copyright and License Information
 */

#include "config.h"

#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <ctype.h>
#include <limits.h>
#include <syslog.h>
#include <libintl.h>
#include <stdio.h>
#include <pwd.h>
#include <errno.h>
#include "pwquality.h"

/* For Translators: "%s%s" could be replaced with "<service> " or "". */
#define PROMPT1 _("New %s%spassword: ")
/* For Translators: "%s%s" could be replaced with "<service> " or "". */
#define PROMPT2 _("Retype new %s%spassword: ")
#define MISTYPED_PASS _("Sorry, passwords do not match.")

/*
 * here, we make a definition for the externally accessible function
 * in this file (this definition is required for static a module
 * but strongly encouraged generally) it is used to instruct the
 * modules include file to define the function prototypes.
 */

#define PAM_SM_PASSWORD

#include <security/pam_modules.h>
#include <security/_pam_macros.h>
#include <security/pam_ext.h>

/* argument parsing */
#define PAM_DEBUG_ARG       0x0001

struct module_options {
        int retry_times;
        int enforce_for_root;
        int local_users_only;
        pwquality_settings_t *pwq;
};

#define CO_RETRY_TIMES  1

#define PATH_PASSWD "/etc/passwd"

static int
_pam_parse (pam_handle_t *pamh, struct module_options *opt,
            int argc, const char **argv)
{
        int ctrl = 0;
        int rv;
        pwquality_settings_t *pwq;
        void *auxerror;
        char buf[PWQ_MAX_ERROR_MESSAGE_LEN];

        pwq = pwquality_default_settings();
        if (pwq == NULL)
                return -1;

        /* just log error here */
        if ((rv=pwquality_read_config(pwq, NULL, &auxerror)) != 0)
                pam_syslog(pamh, LOG_ERR,
                        "Reading pwquality configuration file failed: %s",
                        pwquality_strerror(buf, sizeof(buf), rv, auxerror));

        /* step through arguments */
        for (ctrl = 0; argc-- > 0; ++argv) {
                char *ep = NULL;

                if (!strcmp(*argv, "debug"))
                        ctrl |= PAM_DEBUG_ARG;
                else if (!strncmp(*argv, "type=", 5))
                        pam_set_item (pamh, PAM_AUTHTOK_TYPE, *argv+5);
                else if (!strncmp(*argv, "retry=", 6)) {
                        opt->retry_times = strtol(*argv+6, &ep, 10);
                        if (!ep || (opt->retry_times < 1))
                                opt->retry_times = CO_RETRY_TIMES;
                } else if (!strncmp(*argv, "enforce_for_root", 16)) {
                        opt->enforce_for_root = 1;
                } else if (!strncmp(*argv, "local_users_only", 16)) {
                        opt->local_users_only = 1;
                } else if (!strncmp(*argv, "difignore=", 10)) {
                        /* ignored for compatibility with pam_cracklib */
                } else if (!strncmp(*argv, "reject_username", 15)) {
                        /* ignored for compatibility with pam_cracklib */
                } else if (!strncmp(*argv, "authtok_type", 12)) {
                        /* for pam_get_authtok, ignore */;
                } else if (!strncmp(*argv, "use_authtok", 11)) {
                        /* for pam_get_authtok, ignore */;
                } else if (!strncmp(*argv, "use_first_pass", 14)) {
                        /* for pam_get_authtok, ignore */;
                } else if (!strncmp(*argv, "try_first_pass", 14)) {
                        /* for pam_get_authtok, ignore */;
                } else if (pwquality_set_option(pwq, *argv)) {
                        pam_syslog(pamh, LOG_ERR, 
                                "pam_parse: unknown or broken option; %s", *argv);
                }
         }

         opt->pwq = pwq;

         return ctrl;
}

static int
check_local_user (pam_handle_t *pamh,
                  const char *user)
{
        struct passwd pw, *pwp;
        char buf[4096];
        int found = 0;
        FILE *fp;
        int errn;

        fp = fopen(PATH_PASSWD, "r");
        if (fp == NULL) {
                pam_syslog(pamh, LOG_ERR, "unable to open %s: %s",
                           PATH_PASSWD, pam_strerror(pamh, errno));
                return -1;
        }

        for (;;) {
                errn = fgetpwent_r(fp, &pw, buf, sizeof (buf), &pwp);
                if (errn == ERANGE) {
                        pam_syslog(pamh, LOG_WARNING, "%s contains very long lines; corrupted?",
                                   PATH_PASSWD);
                        /* we can continue here as next call will read further */
                        continue;
                }
                if (errn != 0)
                        break;
                if (strcmp(pwp->pw_name, user) == 0) {
                        found = 1;
                        break;
                }
        }

        fclose (fp);

        if (errn != 0 && errn != ENOENT) {
                pam_syslog(pamh, LOG_ERR, "unable to enumerate local accounts: %s",
                           pam_strerror(pamh, errn));
                return -1;
        } else {
                return found;
        }
}

PAM_EXTERN int
pam_sm_chauthtok(pam_handle_t *pamh, int flags,
                 int argc, const char **argv)
{
        int ctrl;
        struct module_options options;

        memset(&options, 0, sizeof(options));
        options.retry_times = CO_RETRY_TIMES;

        ctrl = _pam_parse(pamh, &options, argc, argv);
        if (ctrl < 0)
                return PAM_BUF_ERR;

        if (flags & PAM_PRELIM_CHECK) {
                /* Check for passwd dictionary
                 * We cannot do that, since the original path is compiled
                 * into the cracklib library and we don't know it.
                 */
                pwquality_free_settings(options.pwq);
                return PAM_SUCCESS;
        } else if (flags & PAM_UPDATE_AUTHTOK) {
                int retval;
                const void *oldtoken;
                const char *user;
                int tries;

                retval = pam_get_user(pamh, &user, NULL);
                if (retval != PAM_SUCCESS || user == NULL) {
                        if (ctrl & PAM_DEBUG_ARG)
                                pam_syslog(pamh, LOG_ERR, "Can not get username");
                        pwquality_free_settings(options.pwq);
                        return PAM_AUTHTOK_ERR;
                }

                retval = pam_get_item(pamh, PAM_OLDAUTHTOK, &oldtoken);
                if (retval != PAM_SUCCESS) {
                        if (ctrl & PAM_DEBUG_ARG)
                                pam_syslog(pamh, LOG_ERR, "Can not get old passwd");
                        oldtoken = NULL;
                }

                tries = 0;
                while (tries < options.retry_times) {
                        void *auxerror;
                        const char *newtoken = NULL;

                        tries++;

                        /* Planned modus operandi:
                         * Get a passwd.
                         * Verify it against libpwquality.
                         * If okay get it a second time.
                         * Check to be the same with the first one.
                         * set PAM_AUTHTOK and return
                         */

                        retval = pam_get_authtok_noverify(pamh, &newtoken, NULL);
                        if (retval != PAM_SUCCESS) {
                                pam_syslog(pamh, LOG_ERR, "pam_get_authtok_noverify returned error: %s",
                                        pam_strerror(pamh, retval));
                                continue;
                        } else if (newtoken == NULL) { /* user aborted password change, quit */
                                pwquality_free_settings(options.pwq);
                                return PAM_AUTHTOK_ERR;
                        }

                        if (options.local_users_only && check_local_user (pamh, user) == 0) {
                                /* skip the check if a non-local user */
                                retval = 0;
                        } else {
                                /* now test this passwd against libpwquality */
                                retval = pwquality_check(options.pwq, newtoken, oldtoken, user, &auxerror);
                        }

                        if (retval < 0) {
                                const char *msg;
                                char buf[PWQ_MAX_ERROR_MESSAGE_LEN];
                                msg = pwquality_strerror(buf, sizeof(buf), retval, auxerror);
                                if (ctrl & PAM_DEBUG_ARG)
                                        pam_syslog(pamh, LOG_DEBUG, "bad password: %s", msg);
                                pam_error(pamh, _("BAD PASSWORD: %s"), msg);

                                if (getuid() || options.enforce_for_root ||
                                    (flags & PAM_CHANGE_EXPIRED_AUTHTOK)) {
                                        pam_set_item(pamh, PAM_AUTHTOK, NULL);
                                        retval = PAM_AUTHTOK_ERR;
                                        continue;
                                }
                        } else {
                                if (ctrl & PAM_DEBUG_ARG)
                                        pam_syslog(pamh, LOG_DEBUG, "password score: %d", retval);
                        }

                        retval = pam_get_authtok_verify(pamh, &newtoken, NULL);
                        if (retval != PAM_SUCCESS) {
                                pam_syslog(pamh, LOG_ERR, "pam_get_authtok_verify returned error: %s",
                                pam_strerror(pamh, retval));
                                pam_set_item(pamh, PAM_AUTHTOK, NULL);
                                continue;
                        } else if (newtoken == NULL) {      /* user aborted password change, quit */
                                pwquality_free_settings(options.pwq);
                                return PAM_AUTHTOK_ERR;
                        }

                        pwquality_free_settings(options.pwq);
                        return PAM_SUCCESS;
                }

                pwquality_free_settings(options.pwq);
                pam_set_item (pamh, PAM_AUTHTOK, NULL);

                /* if we have only one try, we can use the real reason,
                 * else say that there were too many tries. */
                if (options.retry_times > 1)
                        return PAM_MAXTRIES;
                else
                        return retval;
        } else {
                pwquality_free_settings(options.pwq);
                if (ctrl & PAM_DEBUG_ARG)
                        pam_syslog(pamh, LOG_NOTICE, "UNKNOWN flags setting %02X",flags);
        }

        return PAM_SERVICE_ERR;
}



#ifdef PAM_STATIC
/* static module data */
struct pam_module _pam_pwquality_modstruct = {
     "pam_pwquality",
     NULL,
     NULL,
     NULL,
     NULL,
     NULL,
     pam_sm_chauthtok
};
#endif

/*
 * Copyright (c) Cristian Gafton <gafton@redhat.com>, 1996.
 *                                              All rights reserved
 * Copyright (c) Red Hat, Inc, 2011, 2012
 * Copyright (c) Tomas Mraz <tm@t8m.info>, 2011
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, and the entire permission notice in its entirety,
 *    including the disclaimer of warranties.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior
 *    written permission.
 *
 * ALTERNATIVELY, this product may be distributed under the terms of
 * the GNU General Public License version 2 or later, in which case the
 * provisions of the GPL are required INSTEAD OF the above restrictions.
 *
 * THIS SOFTWARE IS PROVIDED `AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The following copyright was appended for the long password support
 * added with the libpam 0.58 release:
 *
 * Modificaton Copyright (c) Philip W. Dalrymple III <pwd@mdtsoft.com>
 *       1997. All rights reserved
 *
 * THE MODIFICATION THAT PROVIDES SUPPORT FOR LONG PASSWORD TYPE CHECKING TO
 * THIS SOFTWARE IS PROVIDED `AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */
